src/backend/joanie/core/management/commands/create_certificates.py

```diff
@@ -0,0 +1,90 @@
+"""Management command to create all available certificates that are not yet existing."""
+import logging
+
+from django.core.management import BaseCommand
+
+from joanie.core import enums, models
+
+logger = logging.getLogger("joanie.core.create_certificates")
+
+
+class Command(BaseCommand):
+    """Browse all courses and their products then create certificate if needed."""
+
+    help = "Browse all courses and their products then create certificate if needed."
+
+    def add_arguments(self, parser):
+        parser.add_argument(
+            "-c",
+            "--course",
+            help="Provide a course code to restrict check to this course",
+        )
+        parser.add_argument(
+            "-p",
+            "--product",
+            help="Provide a product uuid to restrict check to this product",
+        )
+
+    def handle(self, *args, **options):
+        """
+        Retrieve all courses with certifying products then for each product check state
+        of order and their enrollments, if all enrollments has been passed with success
+        we are able to create the related certificate.
+        """
+        # TODO: Improve db queries by using select_related and prefetch_related methods
+        # 1. First retrieve all courses with certifying products
+        course = options["course"]
+        product = options["product"]
+        filters = {"type__in": enums.PRODUCT_TYPE_CERTIFICATE_ALLOWED}
+
+        if course:
+            filters.update({"courses__code": course})
+
+        if product:
+            filters.update({"uid": product})
+
+        products = models.Product.objects.filter(**filters)
+
+        for product in products:
+
+            # Retrieve all active orders related to the course
+            orders = [
+                order
+                for order in product.orders.filter(
+                    is_canceled=False, certificate__isnull=True
+                )
+                if order.state == enums.ORDER_STATE_VALIDATED
+            ]
+
+            certificate_generated = 0
+            for order in orders:
+                graded_courses = order.target_courses.filter(
+                    order_relations__is_graded=True
+                ).order_by("order_relations__position")
+
+                course_runs = models.CourseRun.objects.filter(
+                    course__in=graded_courses, is_gradable=True
+                )
+
+                enrollments = order.enrollments.filter(
+                    course_run__in=course_runs, is_active=True
+                )
+
+                if len(enrollments) < len(graded_courses):
```

Comment:
Can you add a comment about this situation? I assume it means the student still needs to pass some course runs in this course. 
Actually it would be more robust to compare the 2 lists of course runs and not just the number... I'm not sure a student could be subscribed to several course runs of the same course.

Reply:
Yes, from the frontend interface, user is not allowed to enroll to two course runs for the same course. But in fact, from a backend point of view, this is something allowed so I will ensure that all passed enrollments covers all graded courses.